src/main/java/example/GSAJaccardSimilarityMeasure.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.flink.api.common.functions.FlatMapFunction;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.common.functions.ReduceFunction;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.ReduceNeighborsFunction;
+import org.apache.flink.graph.Triplet;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Collector;
+import util.JaccardSimilarityMeasureData;
+
+import java.util.HashMap;
+import java.util.HashSet;
+
+public class GSAJaccardSimilarityMeasure implements ProgramDescription {
+
+	public static void main(String [] args) throws Exception {
+
+		if (!parseParameters(args)) {
+			return;
+		}
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		DataSet<Edge<Long, NullValue>> edges = getEdgesDataSet(env);
+
+		// initialize the vertex values with empty sets
+		Graph<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>, NullValue> graph = Graph.fromDataSet(edges,
+				new MapFunction<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>() {
+
+					@Override
+					public Tuple2<HashSet<Long>, HashMap<Long, Double>> map(Long id) throws Exception {
+						HashSet<Long> neighbors = new HashSet<Long>();
+						neighbors.add(id);
+
+						return new Tuple2<HashSet<Long>, HashMap<Long, Double>>(neighbors,
+								new HashMap<Long, Double>());
+					}
+				}, env);
+
+		// Simulate GSA
+		// Gather: no-op in this case
+		// Sum: create the set of neighbors
+		DataSet<Tuple2<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>> computedNeighbors =
+				graph.reduceOnNeighbors(new GatherNeighbors(), EdgeDirection.ALL);
+
+		// Apply: attach the computed values as vertex values
+		DataSet<Vertex<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>> verticesWithNeighbors =
+				computedNeighbors.map(new MapFunction<Tuple2<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>, Vertex<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>>() {
+
+					@Override
+					public Vertex<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>> map(Tuple2<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>> tuple2) throws Exception {
+						return new Vertex<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>(tuple2.f0, tuple2.f1);
+					}
+				});
+
+		Graph<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>, NullValue> graphWithNeighbors =
+				Graph.fromDataSet(verticesWithNeighbors, edges, env);
+
+		// Scatter: compare neighbors; compute Jaccard
+		DataSet<Vertex<Long, HashMap<Long, Double>>> verticesWithJaccardValues =
+				graphWithNeighbors.getTriplets().flatMap(new ComputeJaccard())
+						.groupBy(0).reduce(new ReduceFunction<Vertex<Long, HashMap<Long, Double>>>() {
```

Comment:
This one is over-complicated as well. On each triplet you will have a pair of neighborhoods. With a single map, you can compare them and output a <srcId, trgId, jaccardValue> tuple, which will be the result.